rosbag2_cpp/src/rosbag2_cpp/clocks/time_controller_clock.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <thread>
+
+#include "rcpputils/thread_safety_annotations.hpp"
+#include "rosbag2_cpp/clocks/time_controller_clock.hpp"
+#include "rosbag2_cpp/types.hpp"
+
+namespace rosbag2_cpp
+{
+
+class TimeControllerClockImpl
+{
+public:
+  /**
+   * Stores an exact time match between a system steady clock and the playback ROS clock.
+   * This snapshot is taken whenever a factor changes such that a new reference is needed,
+   * such as pause, resume, rate change, or jump
+   */
+  struct TimeReference
+  {
+    rcutils_time_point_value_t ros;
+    std::chrono::steady_clock::time_point steady;
+  };
+
+  TimeControllerClockImpl() = default;
+  virtual ~TimeControllerClockImpl() = default;
+
+  template<typename T>
+  rcutils_duration_value_t duration_nanos(const T & duration)
+  {
+    return std::chrono::duration_cast<std::chrono::nanoseconds>(duration).count();
+  }
+
+  rcutils_time_point_value_t steady_to_ros(std::chrono::steady_clock::time_point steady_time)
```

Comment:
@emersonknapp It would be nice to add `REQUIRES(mutex) ` attribute to `steady_to_ros` function. \
https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#getting-started \
However I am not sure if it possible to do with `rclcpp` abstraction layer.